gosling/experimental/data.py

```diff
@@ -0,0 +1,87 @@
+import pathlib
+from typing import Callable, Optional, Union
+
+import gosling.experimental._tilesets as tilesets
+from gosling.experimental._provider import Provider, Resource, TilesetResource
+from gosling.utils.core import _compute_data_hash
+
+
+def _hash_path(path: pathlib.Path):
+    return _compute_data_hash(str(path))
+
+
+class GoslingDataServer:
+    """Backend server for Gosling datasets."""
+
+    def __init__(self) -> None:
+        self._provider: Optional[Provider] = None
+        # We need to keep references to served resources, because the background
+        # server uses weakrefs.
+        self._resources: dict[str, Union[Resource, TilesetResource]] = {}
+
+    @property
+    def port(self):
+        if not self._provider:
+            raise RuntimeError("Server not started.")
+        return self._provider.port
+
+    def reset(self) -> None:
+        if self._provider is not None:
+            self._provider.stop()
+        self._resources = {}
+
+    def __call__(
+        self, data: Union[pathlib.Path, tilesets.Tileset], port: Optional[int] = None
+    ):
+        if self._provider is None:
+            self._provider = Provider(allowed_origins=["*"]).start(port=port)
+
+        if port is not None and port != self._provider.port:
+            self._provider.stop().start(port=port)
+
+        if isinstance(data, tilesets.Tileset):
+            key = "tileset"
+            path = data.filepath
+        else:
+            key = "filepath"
+            path = data
+
+        resource_id = _hash_path(path)
+        if resource_id not in self._resources:
+            self._resources[resource_id] = self._provider.create(**{key: data})
+
+        return self._resources[resource_id].url
+
+    def __rich_repr__(self):
+        yield "resources", self._resources
+        yield "port", self.port
+
+
+data_server = GoslingDataServer()
+
+
+CreateTileset = Callable[[pathlib.Path], tilesets.Tileset]
+
+def _create_loader(type_: str, create_ts: Optional[CreateTileset]= None):
+    def load(url: Union[pathlib.Path, str], **kwargs):
+        """Adds resource to data_server if local file is detected."""
+        fp = pathlib.Path(url)
+        if fp.is_file():
+            data = create_ts(fp) if create_ts else fp
+            url = data_server(data)
+        return dict(type=type_, url=url, **kwargs)
+
+    return load
+
+
+# file resources
+csv = _create_loader("csv")
+json = _create_loader("json")
+bigwig = _create_loader("bigwig")
+
+# tileset resources
+bam = _create_loader("bam", tilesets.bam)
```

Comment:
I did not know that cooler supports bam files, but I am not sure if we need to or should use cooler's tilesets. Does the higlass' python package use this logic? We currently have our own data fetcher for gosling ([here](https://github.com/gosling-lang/gosling.js/tree/master/src/data-fetcher/bam)) like csv, json, and bigwig formats.

Also, using bam in Gosling is recently supported in `v0.9.3`, so we will need to upgrade Gosling.

Q. If I want to upgrade Gosling and test the python package, is there any required command other than `yarn upgrade gosling`?